packages/node/src/indexer/indexer.manager.ts

```diff
@@ -211,136 +293,56 @@ export class IndexerManager {
       return arr;
     }, {} as { [key in typeof keys[number]]: string | boolean | number });
 
-    //blockOffset and genesisHash should only been create once, never update
-    //if blockOffset is changed, will require re-index and re-sync poi.
+    const { chain, genesisHash, specName } = this.apiService.networkMeta;
+
+    // blockOffset and genesisHash should only have been created once, never updated.
+    // If blockOffset is changed, will require re-index and re-sync poi.
     if (!keyValue.blockOffset) {
       const offsetValue = (this.getStartBlockFromDataSources() - 1).toString();
-      await this.storeService.setMetadata('blockOffset', offsetValue);
+      await metadataRepo.upsert({ key: 'blockOffset', value: offsetValue });
+    }
+
+    if (!keyValue.lastProcessedHeight) {
+      if (project) {
+        await metadataRepo.upsert({
+          key: 'lastProcessedHeight',
+          value: project.nextBlockHeight,
```

Comment:
I think this is the scenario when first time subqueryRepo have a nextBlockHeight but metadata table don't have lastProcessdHeight yet, and you try to upsert a value.

But I think this is very dangerous, let me give you an example.

Currently in subqueries table nextBlockHeight is 1000, and you have upsert this 1000 for lastProcessdHeight in meta, what if indexing encounter error at block 1000 and failed, when you re-start index, it will read from meta and start from height 1000+1 = 1001 , so the block 1000 has been skipped.

I rather not upsert the last processed height in here, only update this value when indexd(processed) completed.

Reply:
> I think this is the scenario when first time subqueryRepo have a nextBlockHeight but metadata table don't have lastProcessdHeight yet, and you try to upsert a value.

That's right

> I rather not upsert the last processed height in here, only update this value when indexd(processed) completed.

The problem is that I need a block number to start indexing from and this is based on `lastProcessedHeight`. I'll try to look into how `main` handles this at the moment, I think it's with two different metadata entries.

Reply:
@jiqiang90 
I'm slightly unconvinced that this is a problem. If a block fails to be indexed then the transaction rollsback which should also roll back the `nextBlockHeight` meaning that `nextBlockHeight` is only incremented on successfully indexed blocks.
https://github.com/subquery/subql/blob/a27be97028d64459d7fe895e71bc33198ad688a6/packages/node/src/indexer/indexer.manager.ts#L109-L129

Reply:
I think @jiqiang90 was right about
1) even you want to upsert it to a value it should be 
```
           value: project.nextBlockHeight - 1, 
```
2) for the first block to be processed, nextBlockHeight = startBlock from user's project.yaml, which doesn't mean startBlock - 1 was processed.

Reply:
added a check on the `start` function that attempts to use the find the lastProcessedHeight and then falls back to the `nextBlockHeight` when that doesn't exit (exit otherwise). `lastProcessedHeight init has been removed from the `ensureMetadata` function as well.